### PR TITLE
Friendlier `repr` output for Tables, DataFrames, and Expressions

### PR DESCRIPTION
**Pull request by aaron-siegel**
- Expands `repr` output for tables to include indices, external stores, and view ancestors
- Expands `repr` output for dataframes to print the schema
- Expands `repr` output for columnrefs to print the (one-column) schema
- Introduces `repr` output for all expression types to pretty-print them
- Adds `collect()`, `show()`, `head()`, `tail()`, and `count()` to `ColumnRef` as convenience methods (that are equivalent to `t.select(col_ref).collect()` method, etc)

Also:

- Fixes a bug in which the parameters of a `FunctionCall` were arbitrarily reordered

### CODE REVIEW

**Review by mkornacker — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/384)* status: 0 of 26 files reviewed, 7 unresolved discussions (waiting on @aaron-siegel)

___
*[`tests/test_dataframe.py` line 182 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGOy8hEu3pX9xggzx0:-OCGOy8hEu3pX9xggzx1:b-2j7glt) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/tests/test_dataframe.py#L182)):*
> ```Python
>             .limit(10)
>         )
>         df.describe()
> ```

Should we get rid of describe then?
___
*[`pixeltable/utils/description_helper.py` line 11 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGHyQm8Lr4BcViAalE:-OCGHyQm8Lr4BcViAalF:b-apl4rd) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L11)):*
> ```Python
> class _Descriptor:
>     body: Union[str, pd.DataFrame]
>     show_index: bool
> ```

these only apply to pd.DataFrame body?
___
*[`pixeltable/utils/description_helper.py` line 20 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGHp5zBtlbJyTYaY9n:-OCGHp5zBtlbJyTYaY9o:b-bhuzk9) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L20)):*
> ```Python
>     Helper class for rendering long-form descriptions of Pixeltable objects.
>     """
>     __descriptors: list[_Descriptor]
> ```

comment: why multiple?
___
*[`pixeltable/utils/description_helper.py` line 37 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGIJvYEh4ABdrUTW9K:-OCGIJvYEh4ABdrUTW9L:b-1ao7ot) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L37)):*
> ```Python
>         blocks = [
>             descriptor.body if isinstance(descriptor.body, str)
>             else descriptor.body.to_string(index=descriptor.show_index, header=descriptor.show_header, max_colwidth=49)
> ```

explain 49
___
*[`pixeltable/utils/description_helper.py` line 46 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGIR64BojH8yJZNMs6:-OCGIR64BojH8yJZNMs7:bk4jq77) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L46)):*
> ```Python
>         return '\n'.join(html_blocks)
> 
>     @classmethod
> ```

why classmethod?
___
*[`pixeltable/utils/description_helper.py` line 50 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGIgr10Rm7vTxCUepO:-OCGIgr10Rm7vTxCUepP:b-eurkta) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L50)):*
> ```Python
>         if isinstance(descriptor.body, str):
>             return (
>                 pd.DataFrame([descriptor.body]).style
> ```

What does this do?
___
*[`pixeltable/exprs/column_ref.py` line 135 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGO0v0Ecx2kZPFZUtp:-OCGO0v0Ecx2kZPFZUtq:b-teyzwg) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/exprs/column_ref.py#L135)):*
> ```Python
>         return tbl.select(self)
> 
>     def collect(self) -> 'pxt.dataframe.DataFrameResultSet':
> ```

This is kind of dangerous. It might run for a long time.

File: pixeltable/exprs/column_ref.py [hunk 1]
```diff
@@ -126,18 +128,46 @@ def default_column_name(self) -> Optional[str]:
     def _equals(self, other: ColumnRef) -> bool:
         return self.col == other.col and self.perform_validation == other.perform_validation
 
+    def _df(self) -> 'pxt.dataframe.DataFrame':
+        tbl = catalog.Catalog.get().tbls[self.col.tbl.id]
+        return tbl.select(self)
+
+    def collect(self) -> 'pxt.dataframe.DataFrameResultSet':
```
Comment (mkornacker) [inline pixeltable/exprs/column_ref.py]: Let's remove this, it's really dangerous. If you do frame_view.frame.collect() it'll blow up your memory. show() works just fine for exploration purposes.

I know you can also do this with frame_view.select(frame_view.frame).collect(), but that requires more deliberation.

**Review by mkornacker — CHANGES_REQUESTED** [review]

File: pixeltable/exprs/column_ref.py (hunk 1 above)
Comment (aaron-siegel) [reply pixeltable/exprs/column_ref.py]: Good point, I wasn't really thinking about it when I added it. I've removed `collect()` (but left in `show()`, `head()`, and `tail()`, which are bounded).

**Review by aaron-siegel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/384)* status: 0 of 26 files reviewed, 8 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/exprs/column_ref.py` line 135 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGO0v0Ecx2kZPFZUtp:-OCGuWy8EMLZklY9OlTc:b-896fix) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/exprs/column_ref.py#L135)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

This is kind of dangerous. It might run for a long time.
</blockquote></details>

Done.
___
*[`pixeltable/utils/description_helper.py` line 11 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGHyQm8Lr4BcViAalE:-OCGuiEw04QN6His8hos:b-wthj78) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L11)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

these only apply to pd.DataFrame body?
</blockquote></details>

Correct. I added a clarifying comment
___
*[`pixeltable/utils/description_helper.py` line 20 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGHp5zBtlbJyTYaY9n:-OCGvPSy9XnBARgQQ-r_:b34a5j8) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L20)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

comment: why multiple?
</blockquote></details>

I've elaborated on the docstring comment, which now addresses this.
___
*[`pixeltable/utils/description_helper.py` line 37 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGIJvYEh4ABdrUTW9K:-OCGvSqY8jev2CbXpIs3:bxk4bzq) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L37)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

explain 49
</blockquote></details>

Done. Actually, it was 50, not 49 (I realized this only after tweaking the console display in another way)
___
*[`pixeltable/utils/description_helper.py` line 46 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGIR64BojH8yJZNMs6:-OCGvhsrEOX0w-QI1_UG:b-2oy07n) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L46)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

why classmethod?
</blockquote></details>

It doesn't depend on self, I just forgot to rename self to cls. Fixed.
___
*[`pixeltable/utils/description_helper.py` line 50 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGIgr10Rm7vTxCUepO:-OCGw0-80Laywux1aoOa:b-bizqh8) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/pixeltable/utils/description_helper.py#L50)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

What does this do?
</blockquote></details>

Added comment.
___
*[`tests/test_dataframe.py` line 182 at r1](https://reviewable.io/reviews/pixeltable/pixeltable/384#-OCGOy8hEu3pX9xggzx0:-OCGw1PU3ZPU_YxhGM8j:b53f0oc) ([raw file](https://github.com/pixeltable/pixeltable/blob/aa57802c39306835ae37c71730187bbe3757e711/tests/test_dataframe.py#L182)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Should we get rid of describe then?
</blockquote></details>

Yeah maybe. I'm not sure what purpose it serves. `df.describe()` is exactly equivalent to `display(df)` in a notebook or ipython console (or `print(repr(df))` in a standard python console).

**Review by mkornacker — APPROVED** [review]

**Pull request merged**